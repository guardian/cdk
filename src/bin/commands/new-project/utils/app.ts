import { CodeMaker } from "codemaker";
import type { Imports } from "./imports";
import type { Name } from "./utils";

interface AppBuilderProps {
  imports: Imports;
  appName: Name;
  stack: Name;
  stages: string[];
  outputFile: string;
  outputDir: string;
  comment?: string;
  migrated?: boolean;
}

export class AppBuilder {
  config: AppBuilderProps;

  code: CodeMaker;

  constructor(props: AppBuilderProps) {
    this.config = props;

    this.code = new CodeMaker({ indentationLevel: 2 });
    this.code.closeBlockFormatter = (s?: string): string => s ?? "}";
  }

  async constructCdkFile(): Promise<void> {
    const { comment, outputFile, imports, appName, stack, outputDir, stages } = this.config;

    this.code.openFile(outputFile);
    if (comment) {
      this.code.line(comment);
      this.code.line();
    }

    imports.render(this.code);

    this.code.line("const app = new App();");

    stages.forEach((stage) => {
      this.code.line(
        `new ${appName.pascal}(app, "${appName.pascal}-${stage}", { stack: "${stack.kebab}", stage: "${stage}" });`
      );
    });

    this.code.closeFile(outputFile);
    await this.code.save(outputDir);
  }
}

export const constructApp = async (props: AppBuilderProps): Promise<void> => {
  const builder = new AppBuilder(props);
  await builder.constructCdkFile();
};
