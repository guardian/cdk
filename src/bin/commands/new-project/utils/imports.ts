import type { CodeMaker } from "codemaker";
import type { Name } from "./utils";

interface Import {
  types: string[];
  components: string[];
  basic?: boolean;
}

export class Imports {
  private readonly imports: Record<string, Import>;

  private constructor(imports: Record<string, Import>) {
    this.imports = imports;
  }

  public static newStackImports(addCfnIncludeImports: boolean): Imports {
    return new Imports({
      ...(addCfnIncludeImports && {
        path: {
          types: [],
          components: ["join"],
        },
        "aws-cdk-lib/cloudformation-include": {
          types: [],
          components: ["CfnInclude"],
        },
      }),
      "aws-cdk-lib": {
        types: ["App"],
        components: [],
      },
      "@guardian/cdk/lib/constructs/core": {
        types: ["GuStackProps"],
        components: ["GuStack"],
      },
    });
  }

  public static newAppImports({ kebab, pascal }: Name): Imports {
    return new Imports({
      "aws-cdk-lib": {
        types: [],
        components: ["App"],
      },
      "source-map-support/register": {
        basic: true,
        types: [],
        components: [],
      },
      [`../lib/${kebab}`]: {
        types: [],
        components: [pascal],
      },
    });
  }

  public static newTestImports({ kebab, pascal }: Name): Imports {
    return new Imports({
      "aws-cdk-lib": {
        types: [],
        components: ["App"],
      },
      "aws-cdk-lib/assertions": {
        types: [],
        components: ["Template"],
      },
      [`./${kebab}`]: {
        types: [],
        components: [pascal],
      },
    });
  }

  render(code: CodeMaker): void {
    Object.entries(this.imports)
      // Render "basic" imports before any others, still in alphabetical order
      // Render relative imports after absolute imports
      .sort(([aKey, aImports], [bKey, bImports]) => {
        if (aImports.basic && !bImports.basic) {
          return -1;
        } else if (bImports.basic && !aImports.basic) {
          return 1;
        } else if (aKey.startsWith(".") && !bKey.startsWith(".")) {
          return 1;
        } else if (bKey.startsWith(".") && !aKey.startsWith(".")) {
          return -1;
        } else {
          return aKey.localeCompare(bKey);
        }
      })
      .forEach(([lib, imports]) => {
        imports.basic && code.line(`import "${lib}";`);

        imports.types.length && code.line(`import type { ${imports.types.sort().join(", ")} } from "${lib}";`);

        imports.components.length && code.line(`import { ${imports.components.sort().join(", ")} } from "${lib}";`);
      });
    code.line();
  }
}
