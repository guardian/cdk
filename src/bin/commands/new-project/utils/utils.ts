import camelCase from "lodash.camelcase";
import upperFirst from "lodash.upperfirst";

export const pascalCase = (str: string): string => {
  return upperFirst(camelCase(str));
};

export interface Name {
  kebab: string;
  pascal: string;
}
