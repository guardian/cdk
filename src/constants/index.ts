export * from "./access";
export * from "./context-keys";
export * from "./library-info";
export * from "./regex-pattern";
export * from "./ssm-parameter-paths";
export * from "./tag-keys";
export * from "./tracking-tag";
