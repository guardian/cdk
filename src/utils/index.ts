export * as array from "./array";
export * as exec from "./exec";
export * as git from "./git";
export * as math from "./math";
export * as singleton from "./singleton";
