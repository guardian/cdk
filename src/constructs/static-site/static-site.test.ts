import "@aws-cdk/assert/jest";
import "../../utils/test/jest";
import { SynthUtils } from "@aws-cdk/assert";
import { Stage, StageForInfrastructure } from "../../constants";
import type { SynthedStack } from "../../utils/test";
import { simpleGuStackForTesting, simpleInfraStackForTesting } from "../../utils/test";
import type { GuStaticSiteProps } from "./static-site";
import { GuStaticSite } from "./static-site";

describe("The GuStaticSite pattern", () => {
  const defaultProps: GuStaticSiteProps = {
    app: "my-app",
    domainNameProps: {
      [Stage.CODE]: {
        domainName: "my-app.dev-gutools.co.uk",
      },
      [Stage.PROD]: {
        domainName: "my-app.gutools.co.uk",
      },
    },
  };

  it("should create the correct resources with minimal config", () => {
    const stack = simpleGuStackForTesting();
    new GuStaticSite(stack, "StaticSite", defaultProps);

    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();

    expect(stack).toHaveResourceOfTypeAndLogicalId("AWS::S3::Bucket", /^OriginBucketMyapp.+$/);
    expect(stack).toHaveResourceLike("AWS::CloudFront::Distribution", {
      DistributionConfig: {
        Origins: [
          {
            DomainName: {
              "Fn::GetAtt": ["OriginBucketMyapp924B0FC3", "RegionalDomainName"],
            },
            OriginPath: {
              "Fn::Join": [
                "",
                [
                  "/",
                  {
                    Ref: "Stage",
                  },
                  "/my-app",
                ],
              ],
            },
          },
        ],
      },
    });

    expect(stack).toHaveResourceOfTypeAndLogicalId("AWS::CertificateManager::Certificate", /^CertificateMyapp.+$/);
    expect(stack).toHaveResourceLike("AWS::CloudFront::Distribution", {
      DistributionConfig: {
        ViewerCertificate: {
          AcmCertificateArn: {
            Ref: "CertificateMyappA1452439",
          },
          SslSupportMethod: "sni-only",
        },
      },
    });
  });

  it("should throw when attempting to use an ACM ARN with a hardcoded account id", () => {
    const stack = simpleGuStackForTesting();

    expect(() => {
      new GuStaticSite(stack, "StaticSite", {
        ...defaultProps,
        preExistingCertificateArn: {
          [Stage.CODE]: "arn:aws:acm:us-east-1:000000000000:certificate/abcde-01234-fghij-56789",
          [Stage.PROD]: "arn:aws:acm:us-east-1:000000000000:certificate/01234-abcde-56789-fghij",
        },
      });
    }).toThrowError(
      "Account numbers are considered private information and should not be added to VCS. Use the `account` property on `GuStack` instead."
    );
  });

  it("should throw when provided an ACM ARN outside of the us-east-1 region", () => {
    const stack = simpleGuStackForTesting();

    expect(() => {
      new GuStaticSite(stack, "StaticSite", {
        ...defaultProps,
        preExistingCertificateArn: {
          [Stage.CODE]: `arn:aws:acm:eu-west-1:${stack.account}:certificate/abcde-01234-fghij-56789`,
          [Stage.PROD]: `arn:aws:acm:eu-west-1:${stack.account}:certificate/01234-abcde-56789-fghij`,
        },
      });
    }).toThrowError("CloudFront requires ACM certificates from the us-east-1 region.");
  });

  it("should throw when provided an invalid ACM ARN", () => {
    const stack = simpleInfraStackForTesting();

    expect(() => {
      new GuStaticSite(stack, "StaticSite", {
        ...defaultProps,
        preExistingCertificateArn: {
          [StageForInfrastructure]: `arn:aws:acm:us-east-1:${stack.account}:tls/abcde-01234-fghij-56789`,
        },
      });
    }).toThrowError("Invalid ARN for an ACM resource.");
  });

  it("should not create a new ACM certificate when a us-east-1 certificate ARN is provided", () => {
    const stack = simpleGuStackForTesting();
    new GuStaticSite(stack, "StaticSite", {
      ...defaultProps,
      preExistingCertificateArn: {
        [Stage.CODE]: `arn:aws:acm:us-east-1:${stack.account}:certificate/abcde-01234-fghij-56789`,
        [Stage.PROD]: `arn:aws:acm:us-east-1:${stack.account}:certificate/01234-abcde-56789-fghij`,
      },
    });

    expect(stack).not.toHaveResource("AWS::CertificateManager::Certificate");

    expect(stack).toHaveResourceLike("AWS::CloudFront::Distribution", {
      DistributionConfig: {
        ViewerCertificate: {
          AcmCertificateArn: {
            "Fn::Join": [
              "",
              [
                "arn:aws:acm:us-east-1:",
                {
                  Ref: "AWS::AccountId",
                },
                ":certificate/",
                {
                  "Fn::FindInMap": [
                    "myapp",
                    {
                      Ref: "Stage",
                    },
                    "certificateResourceName",
                  ],
                },
              ],
            ],
          },
          SslSupportMethod: "sni-only",
        },
      },
    });

    const { Mappings } = SynthUtils.toCloudFormation(stack) as SynthedStack;

    expect(Mappings).toMatchObject({
      myapp: {
        CODE: {
          certificateResourceName: "abcde-01234-fghij-56789",
          domainName: "my-app.dev-gutools.co.uk",
        },
        PROD: {
          certificateResourceName: "01234-abcde-56789-fghij",
          domainName: "my-app.gutools.co.uk",
        },
      },
    });
  });

  it("should work with a pre-existing bucket", () => {
    const stack = simpleGuStackForTesting();
    new GuStaticSite(stack, "StaticSite", {
      ...defaultProps,
      preExistingOriginBucketName: "i-already-exist",
    });

    expect(stack).not.toHaveResource("AWS::S3::Bucket");
    expect(stack).not.toHaveResource("AWS::S3::BucketPolicy");

    expect(stack).toHaveResourceLike("AWS::CloudFront::Distribution", {
      DistributionConfig: {
        Origins: [
          {
            DomainName: {
              "Fn::Join": [
                "",
                [
                  "i-already-exist.s3.",
                  {
                    Ref: "AWS::Region",
                  },
                  ".",
                  {
                    Ref: "AWS::URLSuffix",
                  },
                ],
              ],
            },
            OriginPath: {
              "Fn::Join": [
                "",
                [
                  "/",
                  {
                    Ref: "Stage",
                  },
                  "/my-app",
                ],
              ],
            },
          },
        ],
      },
    });
  });

  it("should not create a CloudFormation Mapping when used in a GuStackForInfrastructure", () => {
    const stack = simpleGuStackForTesting();
    new GuStaticSite(stack, "StackSite", defaultProps);
    const json = SynthUtils.toCloudFormation(stack) as SynthedStack;
    expect(json.Mappings).toBeDefined();

    const infraStack = simpleInfraStackForTesting();
    new GuStaticSite(infraStack, "StackSite", {
      ...defaultProps,
      domainNameProps: { [StageForInfrastructure]: { domainName: "my-infra-app.gutools.co.uk" } },
    });
    const infraJson = SynthUtils.toCloudFormation(infraStack) as SynthedStack;
    expect(infraJson.Mappings).toBeUndefined();
  });

  it("should be possible to define multiple in a single stack", () => {
    const stack = simpleGuStackForTesting();
    new GuStaticSite(stack, "Site1", defaultProps);
    new GuStaticSite(stack, "Site2", {
      app: "Site2",
      domainNameProps: {
        [Stage.CODE]: {
          domainName: "site2.dev-gutools.co.uk",
        },
        [Stage.PROD]: {
          domainName: "site2.gutools.co.uk",
        },
      },
      preExistingOriginBucketName: "site2-origin",
    });

    expect(stack).toHaveGuTaggedResource("AWS::CloudFront::Distribution", { appIdentity: { app: defaultProps.app } });
    expect(stack).toHaveGuTaggedResource("AWS::CloudFront::Distribution", { appIdentity: { app: "Site2" } });

    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
  });

  it("should be possible to opt-out of DNS creation", () => {
    const stack = simpleGuStackForTesting();
    new GuStaticSite(stack, "Site1", { ...defaultProps, withoutDns: true });
    expect(stack).toCountResources("Guardian::DNS::RecordSet", 0);
  });

  it("should be possible to create a deterministic bucket name", () => {
    const stack = simpleGuStackForTesting();
    new GuStaticSite(stack, "Site", { ...defaultProps, bucketName: "my-new-bucket" });
    expect(stack).toHaveResourceLike("AWS::S3::Bucket", {
      BucketName: "my-new-bucket",
    });
  });
});
