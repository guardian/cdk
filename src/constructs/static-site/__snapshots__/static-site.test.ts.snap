// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The GuStaticSite pattern should be possible to define multiple in a single stack 1`] = `
Object {
  "Mappings": Object {
    "Site2": Object {
      "CODE": Object {
        "domainName": "site2.dev-gutools.co.uk",
      },
      "PROD": Object {
        "domainName": "site2.gutools.co.uk",
      },
    },
    "myapp": Object {
      "CODE": Object {
        "domainName": "my-app.dev-gutools.co.uk",
      },
      "PROD": Object {
        "domainName": "my-app.gutools.co.uk",
      },
    },
  },
  "Outputs": Object {
    "CloudFrontOAISite2": Object {
      "Description": "Canonical user ID used by CloudFront distribution. This should be granted 's3:GetObject' access to S3 bucket 'site2-origin'.",
      "Value": Object {
        "Fn::GetAtt": Array [
          "CloudFrontOriginAccessIdentitySite23746234B",
          "S3CanonicalUserId",
        ],
      },
    },
    "OriginBucketNameMyapp": Object {
      "Description": "S3 bucket origin for CloudFront.",
      "Value": Object {
        "Ref": "OriginBucketMyapp924B0FC3",
      },
    },
  },
  "Parameters": Object {
    "Stage": Object {
      "AllowedValues": Array [
        "CODE",
        "PROD",
      ],
      "Default": "CODE",
      "Description": "Stage name",
      "Type": "String",
    },
  },
  "Resources": Object {
    "CertificateMyappA1452439": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "DomainName": Object {
          "Fn::FindInMap": Array [
            "myapp",
            Object {
              "Ref": "Stage",
            },
            "domainName",
          ],
        },
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "my-app",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          Object {
            "Key": "Stack",
            "Value": "test-stack",
          },
          Object {
            "Key": "Stage",
            "Value": Object {
              "Ref": "Stage",
            },
          },
        ],
        "ValidationMethod": "DNS",
      },
      "Type": "AWS::CertificateManager::Certificate",
      "UpdateReplacePolicy": "Retain",
    },
    "CertificateSite2260E1692": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "DomainName": Object {
          "Fn::FindInMap": Array [
            "Site2",
            Object {
              "Ref": "Stage",
            },
            "domainName",
          ],
        },
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "Site2",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          Object {
            "Key": "Stack",
            "Value": "test-stack",
          },
          Object {
            "Key": "Stage",
            "Value": Object {
              "Ref": "Stage",
            },
          },
        ],
        "ValidationMethod": "DNS",
      },
      "Type": "AWS::CertificateManager::Certificate",
      "UpdateReplacePolicy": "Retain",
    },
    "CloudFrontOriginAccessIdentityMyapp936D2403": Object {
      "Properties": Object {
        "CloudFrontOriginAccessIdentityConfig": Object {
          "Comment": Object {
            "Fn::Join": Array [
              "",
              Array [
                "Origin Access Identity for my-app ",
                Object {
                  "Ref": "Stage",
                },
              ],
            ],
          },
        },
      },
      "Type": "AWS::CloudFront::CloudFrontOriginAccessIdentity",
    },
    "CloudFrontOriginAccessIdentitySite23746234B": Object {
      "Properties": Object {
        "CloudFrontOriginAccessIdentityConfig": Object {
          "Comment": Object {
            "Fn::Join": Array [
              "",
              Array [
                "Origin Access Identity for Site2 ",
                Object {
                  "Ref": "Stage",
                },
              ],
            ],
          },
        },
      },
      "Type": "AWS::CloudFront::CloudFrontOriginAccessIdentity",
    },
    "DnsRecordMyapp": Object {
      "Properties": Object {
        "Name": Object {
          "Fn::FindInMap": Array [
            "myapp",
            Object {
              "Ref": "Stage",
            },
            "domainName",
          ],
        },
        "RecordType": "CNAME",
        "ResourceRecords": Array [
          Object {
            "Fn::GetAtt": Array [
              "Site1MyappCFDistribution297F4C3E",
              "DomainName",
            ],
          },
        ],
        "Stage": Object {
          "Ref": "Stage",
        },
        "TTL": 3600,
      },
      "Type": "Guardian::DNS::RecordSet",
    },
    "DnsRecordSite2": Object {
      "Properties": Object {
        "Name": Object {
          "Fn::FindInMap": Array [
            "Site2",
            Object {
              "Ref": "Stage",
            },
            "domainName",
          ],
        },
        "RecordType": "CNAME",
        "ResourceRecords": Array [
          Object {
            "Fn::GetAtt": Array [
              "Site2Site2CFDistribution8AFAF8EA",
              "DomainName",
            ],
          },
        ],
        "Stage": Object {
          "Ref": "Stage",
        },
        "TTL": 3600,
      },
      "Type": "Guardian::DNS::RecordSet",
    },
    "OriginBucketMyapp924B0FC3": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "PublicAccessBlockConfiguration": Object {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "my-app",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          Object {
            "Key": "Stack",
            "Value": "test-stack",
          },
          Object {
            "Key": "Stage",
            "Value": Object {
              "Ref": "Stage",
            },
          },
        ],
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "OriginBucketMyappPolicyD18AACC9": Object {
      "Properties": Object {
        "Bucket": Object {
          "Ref": "OriginBucketMyapp924B0FC3",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Principal": Object {
                "CanonicalUser": Object {
                  "Fn::GetAtt": Array [
                    "CloudFrontOriginAccessIdentityMyapp936D2403",
                    "S3CanonicalUserId",
                  ],
                },
              },
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    Object {
                      "Fn::GetAtt": Array [
                        "OriginBucketMyapp924B0FC3",
                        "Arn",
                      ],
                    },
                    "/",
                    Object {
                      "Ref": "Stage",
                    },
                    "/my-app/*",
                  ],
                ],
              },
            },
            Object {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Principal": Object {
                "CanonicalUser": Object {
                  "Fn::GetAtt": Array [
                    "CloudFrontOriginAccessIdentityMyapp936D2403",
                    "S3CanonicalUserId",
                  ],
                },
              },
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    Object {
                      "Fn::GetAtt": Array [
                        "OriginBucketMyapp924B0FC3",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "Site1MyappCFDistribution297F4C3E": Object {
      "Properties": Object {
        "DistributionConfig": Object {
          "Aliases": Array [
            Object {
              "Fn::FindInMap": Array [
                "myapp",
                Object {
                  "Ref": "Stage",
                },
                "domainName",
              ],
            },
          ],
          "Comment": Object {
            "Fn::Join": Array [
              "",
              Array [
                "CloudFront distribution for my-app ",
                Object {
                  "Ref": "Stage",
                },
              ],
            ],
          },
          "DefaultCacheBehavior": Object {
            "AllowedMethods": Array [
              "GET",
              "HEAD",
            ],
            "CachedMethods": Array [
              "GET",
              "HEAD",
            ],
            "Compress": true,
            "ForwardedValues": Object {
              "Cookies": Object {
                "Forward": "none",
              },
              "QueryString": false,
            },
            "TargetOriginId": "origin1",
            "ViewerProtocolPolicy": "redirect-to-https",
          },
          "DefaultRootObject": "index.html",
          "Enabled": true,
          "HttpVersion": "http2",
          "IPV6Enabled": true,
          "Origins": Array [
            Object {
              "ConnectionAttempts": 3,
              "ConnectionTimeout": 10,
              "DomainName": Object {
                "Fn::GetAtt": Array [
                  "OriginBucketMyapp924B0FC3",
                  "RegionalDomainName",
                ],
              },
              "Id": "origin1",
              "OriginPath": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "/",
                    Object {
                      "Ref": "Stage",
                    },
                    "/my-app",
                  ],
                ],
              },
              "S3OriginConfig": Object {
                "OriginAccessIdentity": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "origin-access-identity/cloudfront/",
                      Object {
                        "Ref": "CloudFrontOriginAccessIdentityMyapp936D2403",
                      },
                    ],
                  ],
                },
              },
            },
          ],
          "PriceClass": "PriceClass_100",
          "ViewerCertificate": Object {
            "AcmCertificateArn": Object {
              "Ref": "CertificateMyappA1452439",
            },
            "SslSupportMethod": "sni-only",
          },
        },
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "my-app",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          Object {
            "Key": "Stack",
            "Value": "test-stack",
          },
          Object {
            "Key": "Stage",
            "Value": Object {
              "Ref": "Stage",
            },
          },
        ],
      },
      "Type": "AWS::CloudFront::Distribution",
    },
    "Site2Site2CFDistribution8AFAF8EA": Object {
      "Properties": Object {
        "DistributionConfig": Object {
          "Aliases": Array [
            Object {
              "Fn::FindInMap": Array [
                "Site2",
                Object {
                  "Ref": "Stage",
                },
                "domainName",
              ],
            },
          ],
          "Comment": Object {
            "Fn::Join": Array [
              "",
              Array [
                "CloudFront distribution for Site2 ",
                Object {
                  "Ref": "Stage",
                },
              ],
            ],
          },
          "DefaultCacheBehavior": Object {
            "AllowedMethods": Array [
              "GET",
              "HEAD",
            ],
            "CachedMethods": Array [
              "GET",
              "HEAD",
            ],
            "Compress": true,
            "ForwardedValues": Object {
              "Cookies": Object {
                "Forward": "none",
              },
              "QueryString": false,
            },
            "TargetOriginId": "origin1",
            "ViewerProtocolPolicy": "redirect-to-https",
          },
          "DefaultRootObject": "index.html",
          "Enabled": true,
          "HttpVersion": "http2",
          "IPV6Enabled": true,
          "Origins": Array [
            Object {
              "ConnectionAttempts": 3,
              "ConnectionTimeout": 10,
              "DomainName": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "site2-origin.s3.",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ".",
                    Object {
                      "Ref": "AWS::URLSuffix",
                    },
                  ],
                ],
              },
              "Id": "origin1",
              "OriginPath": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "/",
                    Object {
                      "Ref": "Stage",
                    },
                    "/Site2",
                  ],
                ],
              },
              "S3OriginConfig": Object {
                "OriginAccessIdentity": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "origin-access-identity/cloudfront/",
                      Object {
                        "Ref": "CloudFrontOriginAccessIdentitySite23746234B",
                      },
                    ],
                  ],
                },
              },
            },
          ],
          "PriceClass": "PriceClass_100",
          "ViewerCertificate": Object {
            "AcmCertificateArn": Object {
              "Ref": "CertificateSite2260E1692",
            },
            "SslSupportMethod": "sni-only",
          },
        },
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "Site2",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          Object {
            "Key": "Stack",
            "Value": "test-stack",
          },
          Object {
            "Key": "Stage",
            "Value": Object {
              "Ref": "Stage",
            },
          },
        ],
      },
      "Type": "AWS::CloudFront::Distribution",
    },
  },
}
`;

exports[`The GuStaticSite pattern should create the correct resources with minimal config 1`] = `
Object {
  "Mappings": Object {
    "myapp": Object {
      "CODE": Object {
        "domainName": "my-app.dev-gutools.co.uk",
      },
      "PROD": Object {
        "domainName": "my-app.gutools.co.uk",
      },
    },
  },
  "Outputs": Object {
    "OriginBucketNameMyapp": Object {
      "Description": "S3 bucket origin for CloudFront.",
      "Value": Object {
        "Ref": "OriginBucketMyapp924B0FC3",
      },
    },
  },
  "Parameters": Object {
    "Stage": Object {
      "AllowedValues": Array [
        "CODE",
        "PROD",
      ],
      "Default": "CODE",
      "Description": "Stage name",
      "Type": "String",
    },
  },
  "Resources": Object {
    "CertificateMyappA1452439": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "DomainName": Object {
          "Fn::FindInMap": Array [
            "myapp",
            Object {
              "Ref": "Stage",
            },
            "domainName",
          ],
        },
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "my-app",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          Object {
            "Key": "Stack",
            "Value": "test-stack",
          },
          Object {
            "Key": "Stage",
            "Value": Object {
              "Ref": "Stage",
            },
          },
        ],
        "ValidationMethod": "DNS",
      },
      "Type": "AWS::CertificateManager::Certificate",
      "UpdateReplacePolicy": "Retain",
    },
    "CloudFrontOriginAccessIdentityMyapp936D2403": Object {
      "Properties": Object {
        "CloudFrontOriginAccessIdentityConfig": Object {
          "Comment": Object {
            "Fn::Join": Array [
              "",
              Array [
                "Origin Access Identity for my-app ",
                Object {
                  "Ref": "Stage",
                },
              ],
            ],
          },
        },
      },
      "Type": "AWS::CloudFront::CloudFrontOriginAccessIdentity",
    },
    "DnsRecordMyapp": Object {
      "Properties": Object {
        "Name": Object {
          "Fn::FindInMap": Array [
            "myapp",
            Object {
              "Ref": "Stage",
            },
            "domainName",
          ],
        },
        "RecordType": "CNAME",
        "ResourceRecords": Array [
          Object {
            "Fn::GetAtt": Array [
              "StaticSiteMyappCFDistribution3080CEE2",
              "DomainName",
            ],
          },
        ],
        "Stage": Object {
          "Ref": "Stage",
        },
        "TTL": 3600,
      },
      "Type": "Guardian::DNS::RecordSet",
    },
    "OriginBucketMyapp924B0FC3": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "PublicAccessBlockConfiguration": Object {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "my-app",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          Object {
            "Key": "Stack",
            "Value": "test-stack",
          },
          Object {
            "Key": "Stage",
            "Value": Object {
              "Ref": "Stage",
            },
          },
        ],
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "OriginBucketMyappPolicyD18AACC9": Object {
      "Properties": Object {
        "Bucket": Object {
          "Ref": "OriginBucketMyapp924B0FC3",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Principal": Object {
                "CanonicalUser": Object {
                  "Fn::GetAtt": Array [
                    "CloudFrontOriginAccessIdentityMyapp936D2403",
                    "S3CanonicalUserId",
                  ],
                },
              },
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    Object {
                      "Fn::GetAtt": Array [
                        "OriginBucketMyapp924B0FC3",
                        "Arn",
                      ],
                    },
                    "/",
                    Object {
                      "Ref": "Stage",
                    },
                    "/my-app/*",
                  ],
                ],
              },
            },
            Object {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Principal": Object {
                "CanonicalUser": Object {
                  "Fn::GetAtt": Array [
                    "CloudFrontOriginAccessIdentityMyapp936D2403",
                    "S3CanonicalUserId",
                  ],
                },
              },
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    Object {
                      "Fn::GetAtt": Array [
                        "OriginBucketMyapp924B0FC3",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "StaticSiteMyappCFDistribution3080CEE2": Object {
      "Properties": Object {
        "DistributionConfig": Object {
          "Aliases": Array [
            Object {
              "Fn::FindInMap": Array [
                "myapp",
                Object {
                  "Ref": "Stage",
                },
                "domainName",
              ],
            },
          ],
          "Comment": Object {
            "Fn::Join": Array [
              "",
              Array [
                "CloudFront distribution for my-app ",
                Object {
                  "Ref": "Stage",
                },
              ],
            ],
          },
          "DefaultCacheBehavior": Object {
            "AllowedMethods": Array [
              "GET",
              "HEAD",
            ],
            "CachedMethods": Array [
              "GET",
              "HEAD",
            ],
            "Compress": true,
            "ForwardedValues": Object {
              "Cookies": Object {
                "Forward": "none",
              },
              "QueryString": false,
            },
            "TargetOriginId": "origin1",
            "ViewerProtocolPolicy": "redirect-to-https",
          },
          "DefaultRootObject": "index.html",
          "Enabled": true,
          "HttpVersion": "http2",
          "IPV6Enabled": true,
          "Origins": Array [
            Object {
              "ConnectionAttempts": 3,
              "ConnectionTimeout": 10,
              "DomainName": Object {
                "Fn::GetAtt": Array [
                  "OriginBucketMyapp924B0FC3",
                  "RegionalDomainName",
                ],
              },
              "Id": "origin1",
              "OriginPath": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "/",
                    Object {
                      "Ref": "Stage",
                    },
                    "/my-app",
                  ],
                ],
              },
              "S3OriginConfig": Object {
                "OriginAccessIdentity": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "origin-access-identity/cloudfront/",
                      Object {
                        "Ref": "CloudFrontOriginAccessIdentityMyapp936D2403",
                      },
                    ],
                  ],
                },
              },
            },
          ],
          "PriceClass": "PriceClass_100",
          "ViewerCertificate": Object {
            "AcmCertificateArn": Object {
              "Ref": "CertificateMyappA1452439",
            },
            "SslSupportMethod": "sni-only",
          },
        },
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "my-app",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          Object {
            "Key": "Stack",
            "Value": "test-stack",
          },
          Object {
            "Key": "Stage",
            "Value": Object {
              "Ref": "Stage",
            },
          },
        ],
      },
      "Type": "AWS::CloudFront::Distribution",
    },
  },
}
`;
